include/libopencm3/stm32/h7/rcc.h

```diff
@@ -429,9 +781,64 @@ enum rcc_periph_rst {
 };
 
 #undef _REG_BIT
+/**@}*/
+
 
+/** @defgroup rcc_file RCC peripheral API
+ *
+ * @ingroup peripheral_apis
+ */
 #include <libopencm3/stm32/common/rcc_common_all.h>
 
+BEGIN_DECLS
+
+/**
+ * Setup the base PLLs and clock domains for the STM32H7. This function will
+ * utilize the users input parameters to configure all 3 PLLs, as well as the
+ * core clock domains (namely SYSCLK, CPU, HCLK, AHB, PCLK1-4) with the
+ * specified dividers. Given the dividers, the RCC module will track the
+ * the configured frequency for each of these clock domains.
+ *
+ *   Note: If clock sources, configs, divider, etc. are modified outside of
+ *         this module, the frequency tracking may be undefined.
+ *   Note: Clock tree is fairly complex, see RM0433 Section 7
+ *         for details.
+ * @param[in] config  Input config structure defining desired setup.
+ */
+void rcc_clock_setup_pll(const struct rcc_pll_config *config);
+
+/**
+ * Get the clock rate (in hz) of the specified clock source. There are
+ * numerous clock sources and configurations on the H7, so rates for each
+ * configured peripheral clock are aimed to be discoverd/calculated by this
+ * module such that the user does not need to know how the MCU is configured
+ * in order to utilize a peripheral clock.
+ * @param[in] source  Clock source desired to be fetched.
+ * @return Clock rate in Hz for the specified clock. 0 if undefined or error.
+ */
+uint32_t rcc_get_clock(enum rcc_clock_source source);
+
+/**
+ * Set the clock select for the FDCAN devices.
+ * @param[in] source  Clock source to configure FDCAN kernel clock for.
+ *                    RCC_D2CCIP1R_FDCANSEL_XXX selections above.
+ */
+void rcc_set_fdcan_clksel(uint8_t fdcansel);
+/**
+ * Set the clock select for the SPI 1/2/3 devices.
+ * @param[in] source  Clock source desired to be fetched. Choose from
+ *                    RCC_D2CCIP1R_SPI123_XXX selections above.
+ */
+void rcc_set_spi123_clksel(uint8_t clksel);
+/**
+ * Set the clock select for the SPI 4/5 devices.
+ * @param[in] source  Clock source desired to be fetched. Choose from
+ *                    RCC_D2CCIP1R_SPI45_XXX selections above.
+ */
+void rcc_set_spi45_clksel(uint8_t clksel);
```

Comment:
this is going to explode out too isn't it?  Isn't this what you were adding enums for all these clocks was for?

Reply:
If you look at the comment, it's explicitly stating to use the ```RCC_D2CCIP1R_SPI45_XXX``` selections in the header, which are small values. The function will handle the shifting and what not (which is why the input is clksel and not the reg value). 

I'm open to doing this another way, this was just the way that made sense to me such that the user didn't have to know about the format of the register.

Reply:
No, I mean, you're going to hve functions like this for every single peripheral.  You've only included a couple so far, but following this, they would all get their own function setter.  You've already made an enum for clocks, couldn't something like that be reused here?  (it's pretty much always the same options for the clocks anyway isn't it?)

Reply:
So a few things...all of these can be seen if you reference Table 55 in [RM0433 rev 6.](https://www.st.com/content/ccc/resource/technical/document/reference_manual/group0/c9/a3/76/fa/55/46/45/fa/DM00314099/files/DM00314099.pdf/jcr:content/translations/en.DM00314099.pdf)

1. This is not for every peripheral, as some are grouped, and some are fixed, but yes, there will be a fair few.

1. Unfortunately no, the clksel values are not the same for each device type. In fact, they are even different for the same peripheral type (e.g. SPI1/2/3 and SPI4/5 have completely different selections). See the table, it's a bit of a mess. I can certainly convert these all to enums (code wouldn't change much), but I thought leaving the reg defines would be more appropriate because some folks may want to just make all of the settings themselves without using any of the helpers. I find the helpers useful for readability, modularity, etc.

1. The majority of these peripherals can just be left in their default configuration, and these functions won't need to be used. I thought that leaving this as individual settings would be best from a usage perspective because in most cases, the user probably only sets one or two of them to anything custom, and the rest would get stripped out.

Reply:
so, can you use _either_ rcc_set_peripheral_clk_sel or the specific ones? or do you need to use both?  if you can use either, please drop all the specific ones, it's marginally easier to optimize out, sure, but it's a pile of useless code otherwise